### ISSUE TITLE
[Feature Request] notch macbook support.

### ISSUE THREAD

**Issue by jiengup** [jordanbaird/Ice#1]
Beneficial application!
Do you consider designing for a notch MacBook's built-in screen?
When we have so many menu bar icons, we can't see all the icons using ice's hidden strategy nowadays.
It's like:
On my 27 inch monitor:
<img width="832" alt="image" src="https://github.com/jordanbaird/Ice/assets/46430425/e4bbbc23-6f7a-4c49-ad1d-34ded88533dd">
But on built-in screen:
<img width="1496" alt="image" src="https://github.com/jordanbaird/Ice/assets/46430425/aff06be0-a22f-4a9e-8415-c978bb2d1e39">
because of Apple's genius notch design.

**Comment by jordanbaird** [comment]
This is something that could definitely be supported, but I don't have access to a MacBook with a notch to be able to work on it. I'll add it to the roadmap for the time being, and hopefully it can happen soon.

**Comment by y-u-s-u-f** [comment]
I think it would be cool if we got an option like Bartenders "Bar" - Photo: 
![BartenderBar](https://www.macbartender.com/Bartender5/img/BartenderBar@2x.png)

**Comment by jordanbaird** [comment]
> I think it would be cool if we got an option like Bartenders "Bar" - Photo: ![BartenderBar](https://camo.githubusercontent.com/11f4a11bd4be17578c4dd7d6902fc951b1439e23fc0690e5ecac14c7f48c9886/68747470733a2f2f7777772e6d616362617274656e6465722e636f6d2f42617274656e646572352f696d672f42617274656e6465724261724032782e706e67)

This is also on the roadmap, but can't be implemented until I work out how to implement individual menu bar item movement. Right now, Ice can only move items in sections. I have a semi-working implementation, but it's quite complicated with lots of failure points that need to be addressed.

**Comment by merken** [comment]
I'm waiting for notch support as well. Once this issue gets resolved, I'll try this app out!

**Comment by loranger** [comment]
Notch support would definitely be great: I miss some app icons behind it 😞

**Comment by Xytronix** [comment]
For those additional menu bars, its only possible to add them to the menu bar and then to be opened as far I know. This isn't that fluent, but a solutions for Macbooks with a Notch

**Comment by blu-e-man** [comment]
Now that Bartender is exhibiting strange behavior, notch support from ICE would be great.

**Comment by HaigNalbandian** [comment]
**Please** move this to the top of the backlog. **Many** people want to move from Bartender at the moment, but this is a must-have.

**Comment by yonatankremer** [comment]
the people have spoken

**Comment by nweii** [comment]
Another space-efficient approach for when the notch prevents all extra items from showing could be for those extra items to *replace* the unhidden items when revealed, instead of showing in addition to those items. That creates some extra space to show hidden icons. I use it this way in bartender currently.

**Comment by jordanbaird** [comment]
I hear everyone. This is definitely in the pipeline. The main thing blocking this is that Ice needs to be able to move individual menu bar icons. When an icon is clicked in the secondary bar, it needs to be temporarily moved from whichever section it is in to the visible section so it can show its menu. I have this _mostly_ working, but there are certain apps, such as BetterTouchTool, that block the proper events from being sent to the items. A temporary workaround could be to disable the feature when BetterTouchTool is running.

**Comment by jonasjelonek** [comment]
Is there anything we can support you with in this, besides the financial support?
This is currently the only thing that unwillingly prevents me from moving from Bartender to Ice, since I also have the Notch.

Maybe an alpha/beta release would be something? Since you said it's mostly working we could test it, and discover and sort out remaining issues.

**Comment by merken** [comment]
I would also like to volunteer as beta tester if that helps. MBP 14 with notch.
Would it not be as simple as providing a spacer at the coordinates of the notch? Are there any API's to leverage here?

**Comment by MoKhajavi75** [comment]
I can help in beta testing too. MBP 14 (M3 - Sonoma 14.5)

**Comment by HaigNalbandian** [comment]
Also willing to test on my notched 14" MB Air and on a 27" 1440p monitor

**Comment by gingerbeardman** [comment]
> Would it not be as simple as providing a spacer at the coordinates of the notch? Are there any API's to leverage here?

My thoughts exactly. If so, we would need to figure out if the current display has a notch (hopefully as simple as an api call) and how big it is (are they all the same?)

**Comment by fifafu** [comment]
@jordanbaird developer of BetterTouchTool here, what kind of events are you seeing blocked by BTT? Would be happy to fix that.

**Comment by kud** [comment]
It could be amazing to have this feat. I want to get rid of Bartender asap. 🙏

**Comment by jordanbaird** [comment]
@fifafu Thank you for the help! I haven't tracked down exactly which of these it is, but it's one (or all) of mouse down, mouse dragged, and mouse up. It results in an event tap that listens for these events timing out. Would you be able to email me at jordanbaird.dev@gmail.com so we can talk in more detail about what's going on?

**Comment by fifafu** [comment]
Sure I'll ping you!
But isn't the general problem that you can only target items that are "on-screen" with these events? I can easily programmatically move around the items that are visible, but the hidden ones that are pushed off screen by setting the status item length to a high value, these I'm unable to access using standard events.

**Comment by jordanbaird** [comment]
@fifafu I've found a workaround for that. But yes, that was a problem early on.

**Comment by jordanbaird** [comment]
Good news -- @fifafu and I were able to solve the remaining problems with individual item movement. All that's left now is to implement the secondary bar

**Comment by MoKhajavi75** [comment]
We really appreciate your efforts @jordanbaird & @fifafu ✌🏻

**Comment by dwt** [comment]
@jordanbaird as a fellow developer (though mostly python these days) I'd be glad to help test and send logs or do experiments. Yes my ObjC/Swift is rusty, but I should be able to do some basic experiments.

**Comment by ThisIsMissEm** [comment]
Just gonna through this out there: if you change your screen size to a non-standard size the stupid top notch goes away, and you can see everything.

i previously tried to use TopNotch to remove it, and that stopped working when I upgraded to M3 / Sonoma, but I found that workaround after a lot of googling

**Comment by dwt** [comment]
I'm using the tool [zNotch](https://github.com/zkondor/znotch) temporary hide/show the notch, together with this handy shell script:

```shell
$ cat bin/toggle-notch
#!/usr/bin/env sh
open 'xyz.kondor.znotch://v1/manage?action=toggle'
```
That gives me access to all my menu items in a pinch until ice has better support.

**Comment by jordanbaird** [comment]
I have something that I think works pretty well. I'll attach a build for anyone who wants to try it out. In addition to the secondary bar, you can now rearrange your menu bar items directly within Ice's interface! I would appreciate anyone's feedback! :)

[Ice.zip](https://github.com/user-attachments/files/15818709/Ice.zip)

### Some known issues:

1. The purple screen capture icon appears in the menu bar quite often. This is because Ice needs to capture images of the individual menu bar items to display in its interface (these and portion of the desktop background below the menu bar are the only things Ice ever captures). I'll be optimizing this.
2. Rapidly switching spaces can cause the icons to disappear from the secondary bar. A workaround if this happens is to switch to another space, wait for a few seconds, then switch back.
3. Menu bar items that are temporarily shown by clicking them in the secondary bar are currently rehidden after a fixed amount of time. This will be changed to use a dynamic check, similar to Ice's smart autohide.
4. The displayed menu bar items (both in the secondary bar and the arranging interface) can sometimes get out of sync with the actual items in the menu bar.

### Things to test:

1. We need to make sure that the current method of hiding/showing items works the same as it did. I had to rework some of the internals to get the two methods to play nice with each other.
2. When an external display is connected, the secondary bar should appear on the currently active display. I have not tested this, so it would be great to have confirmation that it works (or doesn't).

### Other things:

1. Is the "Secondary Bar" name okay? Or would "Ice Bar" be better?

**Comment by fifafu** [comment]
Nice!
Fyi, to get rid of the purple screen recording icon you can start the app like this:
```
screen -dmS BetterTouchTool /Applications/Ice.app/Contents/MacOS/Ice
```

**Comment by luckman212** [comment]
I vote for "Ice Bar" - sounds cool! 😎

**Comment by jonasjelonek** [comment]
@jordanbaird I just tested your build. My device is MacBook Pro 14 M1 Pro with macOS 14.5

Works good so far, here my experience to you concrete testing points:
1. Seems to work, but someone else should verify this. At least I didn't noticed anything that looks strange to me or doesn't work.
2. This doesn't seem to work, at least not in my setup. I've attached a second monitor and if I click on the dot in the menu bar on my second monitor, the second bar appears on my main MacBook display.

Apart from that I just noticed: At first start the Ice icons/separators seem to be placed at the left end of the menu bar. In my case, where I already have too many icons to fit there because of the notch, I didn't found a proper way then to move the Ice icons the hide excess items. IIRC, Bartender by default includes most of the items in the menu bar at first start, so setting this up is much easier. Do you get the issue and is that solvable similar to how Bartender does this?

**Comment by wombley** [comment]
Thanks for the update! Testing on a Macbook Pro 14" M1 Pro, 14.5 - some issues:

With appearance: split, the bar is now cut off (this showed up as soon as I launched the new app, before I enabled secondary bar, so may be related to some other change). Playing around with this has it cut off in various locations (especially as screen recording icon comes and goes). You can click on the hidden section and it's clear the hidden items are receiving the click as their dropdowns show - even ones further along than you'd expect if it's just the cutoff for the rounded corners.

<img width="919" alt="Screenshot 2024-06-13 at 13 17 50" src="https://github.com/jordanbaird/Ice/assets/252025/24f9ba08-f51c-441e-b3f1-6173113f31bc">

<img width="161" alt="Screenshot 2024-06-13 at 13 18 33" src="https://github.com/jordanbaird/Ice/assets/252025/32e1c5e4-e37d-4fc5-b02e-e683be3248c8">

<img width="669" alt="Screenshot 2024-06-13 at 13 34 12" src="https://github.com/jordanbaird/Ice/assets/252025/3e381212-ad25-4e87-8e86-7cb4ab92fac8">



If secondary bar is enabled when the hidden items have already been shown / are currently showing, they're then duplicated in both locations - and stay visible in the menu bar even when the show/hide button is repeatedly pressed (note using the ellipsis as separator button).

<img width="691" alt="Screenshot 2024-06-13 at 13 24 49" src="https://github.com/jordanbaird/Ice/assets/252025/23c2d702-26c8-4450-93a5-d12debafc009">



(This may be outside your current objective, and is minor besides) - with bartender the secondary bar showed up below the show/hide button, with Ice it seems to always show up on the far right.

Thanks!

**Comment by gingerbeardman** [comment]
Icons are always showing as dark in the Ice Bar for me. light mode or dark mode

![Screen shot 2024-06-13 at 16 25 17](https://github.com/jordanbaird/Ice/assets/49612/3f9c8fd9-0be9-4d61-89b8-8a0e516f904d)

**Comment by lyrst** [comment]
> When an external display is connected, the secondary bar should appear on the currently active display. I have not tested this, so it would be great to have confirmation that it works (or doesn't).

This works not for me, so same behavior as described by [jonasjelonek](https://github.com/jonasjelonek).

In addition to that, the 2nd bar has a strange ratio when I open it from the external screen, then it shown like this on the macbook screen
<img width="168" alt="image" src="https://github.com/jordanbaird/Ice/assets/57271116/a4e1677a-0229-41ca-a323-55420c1bb0cf">

**Comment by jordanbaird** [comment]
New beta:

[Ice.zip](https://github.com/user-attachments/files/15829629/Ice.zip)

I'm not in a spot where I can test on an external display at the moment, so bear that in mind.

This build includes the following changes:

- Fixed inconsistencies when toggling between the secondary panel and standard mode
- Fixed secondary bar not appearing on external displays (hopefully)
- Fixed secondary bar not updating with current items
- Items in the secondary bar should now be displayed with the correct aspect ratio (hopefully)
- Allow secondary bar to be moved via drag
- Don't show sections when Command+dragging if using the secondary bar
- Cache item images to prevent excessive screen capture use
- Rehide temporarily shown items when showing the secondary bar
- Showing item menus is now more reliable
- Wait for item menus to close before rehiding

Also, I think I will be going with "Ice Bar" for the name.

**Comment by jordanbaird** [comment]
@scp Thanks for the info! Does the layout interface (`Menu Bar Items` in Ice's settings) also display items in an incorrect aspect ratio? When opening the settings window on both screens?

**Comment by BossElijah** [comment]
Hey @jordanbaird thanks for the beta package. Do you have any idea when this will be released for real?

**Comment by rockyzhang24** [comment]
The author is already adapting very quickly, so don't rush. Based on the discussion, it should be released soon. @jordanbaird Thank you so much for your hard work in developing and maintaining this. We're really looking forward to it. 🫡

**Comment by lint2015** [comment]
I tried the latest beta posted on a 27" iMac. I'm using the Chevron option as the Ice icon.

The Menu Bar items section of the settings doesn't really work. The Hidden and Always-Hidden sections are both blank. When I drag icons from the Visible section to the Hidden section, they either disappear or they appear in the Hidden section briefly before disappearing. This also causes the large "Hidden" chevron to move to unexpected places on the menu bar (once it moved to the left of the smaller "Always-Hidden" chevron).

When I fix things my Cmd+drag on the actual menu bar, all the icons that should be in the Hidden section appear properly under Menu Bar Items, but after some time, that section becomes blank again.

When I switch on "Use secondary bar" I get a blank secondary bar that's looks like it's the correct width for all of my Hidden icons, but without the icons there.

Happy to help debug if required.

**Comment by Xytronix** [comment]
How does a setting for moving the application settings to the secondary bar sound? 

Is this something doable?

**Comment by jonasjelonek** [comment]
It now works that the secondary bar is displayed on the external display in the correct cases. And I can also properly drag the secondary bar. So that seems to be fixed.

But remaining issues I noticed:
- the secondary bar still has some aspect ratio issues. If I start Ice and first open the secondary bar on my main screen, it is properly scaled there but then not on my external display. But if I start Ice and first open the bar on the external display, it is properly scaled there but not on my main screen. This remains until I restart Ice, and only shows properly on one of both screens. 
- after dragging the secondary bar, the position doesn't seem to be remembered? So after closing and opening it again, it will appear again at the most right of the screen.
- dynamic icon color: the secondary bar doesn't seem to adopt dynamically if the icon color in the menubar changes from black to white in case I change my wallpaper. So menubar items become white, but the items in the secondary bar remain black. This is resolved by restarting Ice.

**Comment by gingerbeardman** [comment]
> New beta:

Icons in secondary bar are still wrong colour for me

![image](https://github.com/jordanbaird/Ice/assets/49612/acda1c14-7132-4add-8021-1307fbe34df0)

![Screen shot 2024-06-14 at 15 38 47](https://github.com/jordanbaird/Ice/assets/49612/79f7fa72-a50d-457d-9928-d12c734fd93d)

**Comment by Xytronix** [comment]
<img width="372" alt="image" src="https://github.com/jordanbaird/Ice/assets/32957125/87e38484-996b-4475-92b9-cda06373f3e1">
Instead of showing this popup when trying to move the clock, make the clock non movable or show it only on the first time

**Comment by wzxu** [comment]
Thanks so much for implementing this! Been using iBar recently but finally I can switch away.

A few observations compared to iBar that Ice could replicate to make a better experience:
1. Secondary bar should show below the cursor instead of always at the far right.
2. When clicking an icon on the secondary bar, in addition to showing that icon on the menubar, it should also already been focused/activated, so we don't have to click on that icon again on the menubar.

**Comment by dkarter** [comment]
Super excited for this, thank you for implementing it!

Here are a couple of bugs I noticed:

1. I'm also seeing the split bar cutting off icons, an easy way to test this would be using the built in battery indicator and enabling show battery percentage in settings:
![CleanShot 2024-06-14 at 11 36 31@2x](https://github.com/jordanbaird/Ice/assets/551858/05963b8f-646f-4be7-9dee-54b8a6cf8456)

2. Also noticed some items getting warped and icons that are dynamically changing are not updating in the secondary bar:

https://github.com/jordanbaird/Ice/assets/551858/01299810-5294-40b7-8034-daebaf1570dd

**Comment by holdit** [comment]
Suggestion: centre the bar under the Ice button.

For those using a mouse to see the hidden icons, it's better if the bar is right under Ice's button than to have to go all the way to the right to interact with the bar (keep in mind that the Ice icon will be used in different positions and the bar will have different widths).

Probably not something that must be implemented right away, but it's an idea for the future. Maybe it could be enabled/disabled with a setting? I assume some prefer if it's aligned to the right.

**Comment by 0rly0wl** [comment]
Thanks a lot for making this, your work is really appreciated!

When Ice is a more finished product, will you consider submitting it to Mac App Store? 
I would gladly pay (one time payment) for the app if it's on there, as would many other people

**Comment by notmyalien** [comment]
@jordanbaird I'm Running this beta on Mac Sequoia beta 1 with minimal issues. Only thing I'm encountering is sometimes it takes multiple times via the settings menu to get an app icon to 'stick' where I want it, in the permanently visible section of the menu bar.

**Comment by wzxu** [comment]
I noticed that the secondary bar won't show up when in fullscreen.
I wonder if it's not shown topmost but as a normal window so getting overlapped, or something.

**Comment by sweetppro** [comment]
@jordanbaird while running v0.10.0-beta-3...
in addition to centering of the icebar, icebar background color and having to also click on menu items after selecting them in the icebar - the Ice menu item highlight color flickers upon clicking. It would be great to remove the flicker

otherwise, nice work
:)

**Comment by TheLinuxGuy** [comment]
Really excited for this FR to land soon so that I can run away from Bartenderapp... great work @jordanbaird

**Comment by fleytman** [comment]
I really look forward to this feature, it will allow you to go to your open source solution.
While people are waiting, it is not necessary to buy Bartenders, there are free alternatives
Previously mentioned [ibar](https://apps.apple.com/cn/app/ibar-强大的菜单栏图标管理工具/id6443843900), in addition to the panel can reduce the size of the icon, but you can also [without it](https://www.jessesquires.com/blog/2023/12/16/macbook-notch-and-menu-bar-fixes/)
```bash
defaults -currentHost write -globalDomain NSStatusItemSpacing -int 12
defaults -currentHost write -globalDomain NSStatusItemSelectionPadding -int 8
```
and relogin user (or reboot system)

There is also a [barbee](https://apps.apple.com/us/app/barbee-hide-menu-bar-items/id1548711022), but it works worse for me 
Function similar to [znotch](https://github.com/zkondor/znotch) and hiding icons like ice is available in [only switch](https://github.com/jacklandrin/OnlySwitch) (as well as convenient tweaks)

**Comment by Chamiu** [comment]
0.10.0-beta 3 is fantastic.
Users can explicitly categorise Menubar items according to their preferences !
 @jordanbaird

**Comment by billchase2** [comment]
@jordanbaird On the latest beta, I'm only seeing icons in the "Visible" section in the settings under Menu Bar Items and none are appearing under the "Hidden" and "Always-Hidden" sections. The menubar itself is fine, it's just in the settings where no icons are visible in those two sections.

**Comment by Chamiu** [comment]
Ice 0.10.b3 on macOS Sonoma.5 on MacBook Air Late 2020.

All visible.
![0 10 0-b3 macOS Sonoma 5](https://github.com/jordanbaird/Ice/assets/2077600/732e7a3a-70c9-4027-86db-646d630e6adb)

Is it environmentally dependent ?

**Comment by billchase2** [comment]
I also just ran into an issue with overlapping icons:
<img width="108" alt="Finder Menu Icon" src="https://github.com/jordanbaird/Ice/assets/16213205/cc06f985-67e5-455c-940d-aaaeccfe37fb">

**Comment by jordanbaird** [comment]
@billchase2 Can you elaborate on that a little more? Are they always overlapping? Or just under certain conditions?

**Comment by dudeofawesome** [comment]
The beta has been working very well for me!

I have found a small bug though: when I click on the clock to open the Notification Center, the secondary bar shows up.

https://github.com/jordanbaird/Ice/assets/1683595/a7cc83ac-98d2-4a75-a356-cae1867b290a

Here's my Ice settings, in case that helps:
<img width="1012" alt="image" src="https://github.com/jordanbaird/Ice/assets/1683595/7dbc0d04-6c56-43ee-a338-e9a008dc1cdb">

**Comment by billchase2** [comment]
> @billchase2 Can you elaborate on that a little more? Are they always overlapping? Or just under certain conditions?

It looks like it just while opening specific apps. In this instance, it's "Keep it Shot." I tried other apps with menubar icons, like NextDNS, and it's not doing it.

**Comment by jordanbaird** [comment]
@billchase2 This looks like a bug with Keep it Shot. Whatever it is, it's not a real menu bar item. You can see here that it exists in addition to the real one:

<img width="1440" alt="image" src="https://github.com/jordanbaird/Ice/assets/90936861/71a0c1bc-6e60-4850-913d-855591632970">

It doesn't look like Keep it Shot is a native macOS app. It's probably doing something hacky to get its menu bar item to work. It appears like this even when Ice isn't running, so it can't be a bug with Ice. I would contact the developers of Keep it Shot.

**Comment by billchase2** [comment]
> @billchase2 This looks like a bug with Keep it Shot. Whatever it is, it's not a real menu bar item. You can see here that it exists in addition to the real one:
> 
> <img alt="image" width="1440" src="https://private-user-images.githubusercontent.com/90936861/340516532-71a0c1bc-6e60-4850-913d-855591632970.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg3MTM4NzMsIm5iZiI6MTcxODcxMzU3MywicGF0aCI6Ii85MDkzNjg2MS8zNDA1MTY1MzItNzFhMGMxYmMtNmU2MC00ODUwLTkxM2QtODU1NTkxNjMyOTcwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjE4VDEyMjYxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE3NWZmYzlhZWI1NTJlYjI1NjBiNmY5MGRmMzZhZTIwNWFiNDFkZjczNmNjZjUxMGE4MWE1ODBjMzgzN2UxMGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.PsVC_7_JHqt2RMgBEhQ7qXm_yqTDKXhcsZSXNFOCEmM">
> It doesn't look like Keep it Shot is a native macOS app. It's probably doing something hacky to get its menu bar item to work. It appears like this even when Ice isn't running, so it can't be a bug with Ice. I would contact the developers of Keep it Shot.

Ah, got it. Apologies for the mix-up and thanks for checking into it! I'll let them know.

**Comment by jordanbaird** [comment]
Here's the latest beta.

[Ice.zip](https://github.com/user-attachments/files/15890325/Ice.zip)

I'm hoping to have worked out most of the problems mentioned here, but I guess we'll see.

Items should now render in the correct aspect ratio (for real this time).

The Ice Bar now appears centered below the Ice icon. Before the official release, I'll be adding an option to specifically choose where it appears (under the Ice icon, under the cursor, left of screen, right of screen). A little Easter Egg that I haven't figured out the best way to present in the UI is the ability to pin the Ice Bar to a specific location on-screen. Currently, you can access this feature by right clicking the bar. While writing this, I tested the pin feature just to make sure it worked and realized it only pins to the primary screen, so that will be fixed in the next beta.

I reverted some of the code that I think caused the issue with rendering the menu bar appearance. I think this will also fix the problem with showing the notification center.

Colors should render more consistently. If they don't, a temporary workaround is to try switching spaces.

The Ice Bar now appears in fullscreen.

Ice should now more reliably click items after temporarily showing them. I also experimented with using Accessibility to perform the "press" action, but some apps don't appear to implement that action. I even had one crash. For now simulating a click on the menu bar item seems like the best option.

There are lots of other smaller fixes, but I think these are the main ones. Let me know how this beta works for you!

**Comment by jonasjelonek** [comment]
Most of the issues that were mentioned before are now gone and it is working great imho.
Only thing I'm experiencing is the clicking of the items in the secondary bar, which doesn't seem to work in my case, for no icon. So I have to click the icon first in the secondary bar and then again after it appeared on the main menu bar.

The pin feature works, but is behaving a bit strange when using two screens. When I pin the bar on my main display, right below the menu bar and then open it on my external display, it is shown in the lower right corner.

**Comment by T-Specht** [comment]
For me, clicking on icons in the secondary bar does not work sadly. All the menu bar items appear correctly in the secondary bar, however, once I click on an icon, my mouse cursor disappears for about a second, the secondary bar closes, the cursor reappears but the menu bar icon I clicked on never shows up in the main menu bar. This happens on the internal display and as well on external monitors.

This was the case in beta 3 and is still present now in beta 4. I am using a MacBook Pro 14" M1 Max. Thank you for your work!

**Comment by Xytronix** [comment]
weird bugs happening when new menu bar items get added:
<img width="561" alt="image" src="https://github.com/jordanbaird/Ice/assets/32957125/aff68937-8b59-4a22-9fc8-b3c6fd554b01">

**Comment by jordanbaird** [comment]
@T-Specht What happens when you try to move items in Ice's settings in the "Menu Bar Items" section? Both on-screen and off-screen items. Also, did you happen to use Bartender? Ice uses a very similar method of moving individual items. I'd like to know if Bartender works on your machine.

**Comment by jordanbaird** [comment]
@Xytronix Strange. Is it happening on an external display? Or the main one? Or both?

**Comment by T-Specht** [comment]
> What happens when you try to move items in Ice's settings in the "Menu Bar Items" section? Both on-screen and off-screen items. Also, did you happen to use Bartender? Ice uses a very similar method of moving individual items. I'd like to know if Bartender works on your machine.

Moving the items in settings sometimes works, but especially for off screen items I often get "operation timed" out errors. The most reliable way seems to be moving the items directly in the menubar for me. And yes, I have used Bartender with the secondary menu bar before switching to Ice and never had any problems.

**Comment by jonasjelonek** [comment]
> @jonasjelonek I'll look into the click issue some more. Maybe it would be a good idea to try the click, and if it fails (if we can somehow check for that), then try the Accessibility method?

I guess it's at least worth trying. Strange, I guess it works on your side, don't know why not on my side. Would be interesting how Bartender does this since it just works there.

**Comment by holdit** [comment]
Getting "Operation timed out" errors almost every time I move any icon. Using the latest beta:

<img width="867" alt="time out" src="https://github.com/jordanbaird/Ice/assets/23439751/dde72c15-8c34-49d9-920a-906c262e7b75">

**Comment by vr8hub** [comment]
10 beta 4 on 16" MBP running Sonoma 14.4.1.
As @gingerbeardman mentioned [above](https://github.com/jordanbaird/Ice/issues/1#issuecomment-2166001281), the Ice Bar is showing up with a dark background and dark icons. I don't use dark mode anywhere, so it's not coming from a system setting.

I also have the same problems as [here](https://github.com/jordanbaird/Ice/issues/1#issuecomment-2176807893), i.e. clicking on icons in the Ice Bar doesn't do anything, so I still can't click to open those apps.

Regardless, fantastic work on this beta, thank you so much for all of your efforts!

I believe it's been mentioned already, but for the future, it would be great if there was an option to have the hidden icons swap out with the visible ones (first click) and then swap back (second click), rather than use a separate Ice Bar. Bartender has been able to do this for a while, and it's how I've always used it.

**Comment by lint2015** [comment]
Thanks for the latest beta. There are some improvements and one regression for me. For the below issues I've deleted the plist file to start fresh with the same results:

- Dragging icons between different sections in Preferences > Menu Bar Items appears to work reliably now, and moving icons there is reflected on the actual menu bar
- However, the Hidden and Always-Hidden sections are still blank if those sections are hidden on the menu bar itself clicking "Menu Bar Items" on the sidebar of Preferences
- When enabling "Use Ice Bar," the bar was the correct size but blank in beta 3. Now in beta 4, the Ice Bar appears to be zero size with just the rounded corners. I wonder if the Ice Bar showing up blank in beta 3 and zero size in beta 4 are both related to the second point above where icons in the Hidden section don't appear to be captured by Ice unless they're already visible at the time

Please see videos below. I will also test on my MBP with notch soon. Thanks again!

https://github.com/jordanbaird/Ice/assets/39164595/e953156e-f15a-43a6-adc9-15cbc4d8e4f5

**Comment by lint2015** [comment]
Video of zero-sized Ice Bar:
https://github.com/jordanbaird/Ice/assets/39164595/580ceaa3-9975-40fc-be70-982caffdb072

**Comment by T-Specht** [comment]
> @T-Specht Are you using any apps that deal with system events, especially mouse events? There could be something that's interfering with the events that Ice sends.

@jordanbaird Thank you for that hint! Turns out the problem occurs when [Magnet (window manager with snapping)](https://magnet.crowdcafe.com/) is running. Once I quit Magnet, everything works as expected. Out of curiosity, I also tried to use Rectangle as a replacement and that seemed to work ok. Any ideas what might cause that behavior?

**Comment by Xytronix** [comment]
Bug: When using a fullscreen app where no menu bar is shown, Ice Bar still appears if you click or hover at the place where the Ice Settings Icon should be on the menu bar.

**Comment by Chamiu** [comment]
ver.0.10.0b3は、どうしても1PasswordをAlways-Hidden Sectionに入れたいようだ。

1. open Ice Menubar Items of Setting
2. 1Password move from Always-Hidden to Hidden
3. macOS reboot
4. noting 1Password in hidden bar
5. -> 1

**Comment by jordanbaird** [comment]
@T-Specht I'm not totally sure what could be the cause, but I have found a temporary workaround for the time being. It unfortnuately makes movement a little slower all around, but at least it's more consistent. We probably won't be able to do menu bar item groups or profiles until a real solution is found, as both require extremely fast movement. But for the current set of features, this should work fine.

**Comment by jordanbaird** [comment]
@Chamiu That is a known bug, and not related to this issue. See #6. Please keep the discussion here related to notch support.

**Comment by Xytronix** [comment]
So since macos 15 I get dozens of popups like this:
<img width="409" alt="image" src="https://github.com/jordanbaird/Ice/assets/32957125/ce7c91f8-f981-4915-93c1-dccc8afa0d4d">

This is very annoying, anything that could be run to keep it to a minimum?

**Comment by Xytronix** [comment]
When changing resolution with Better Touch Tool Ice Bar gets displayed very big:
![image](https://github.com/jordanbaird/Ice/assets/32957125/d8cca018-b6cb-4fc0-bb49-b605ab2687c4)

**Comment by jordanbaird** [comment]
@Xytronix that is annoying. Hopefully it's just an issue with the macOS 15 beta. It's not really specific to the Ice Bar though. Could you open a separate issue for it?

**Comment by Xytronix** [comment]
@jordanbaird did so.

The external monitor implementation is kinda broken.

Sometimes Ice Bar keeps up popping small, big, 1-2 icons big rest small and sometimes appears on the main display despite clicking on the external monitor ice settings icon.

**Comment by Xytronix** [comment]
<img width="1138" alt="image" src="https://github.com/jordanbaird/Ice/assets/32957125/7e6ad6e4-e404-4d94-a341-ab1645c55069">
Bug: Wrong color detecting. Or is it choosing the application color as a background for ice bar?
Happens after resizing application.

**Comment by jordanbaird** [comment]
@Xytronix it should be averaging the color of the desktop background below the menu bar. Looks like you're in fullscreen, though, which I didn't account for.

**Comment by alecgorge-openai** [comment]
> 1. Is the "Secondary Bar" name okay? Or would "Ice Bar" be better?

how about Ice Tray :)

thanks for the cool app!

**Comment by yoelsusanto** [comment]
Hi @jordanbaird thanks for your hard work on this. I am very interested in this feature because I can't find it other than in Bartender.

I saw that you provided several beta builds on this thread, do you plan on creating a WIP branch too? I think this would allow more people to tinker and possible test some things out for ourselves 😃

**Comment by fleytman** [comment]
On latest beta, i can't use on second  bar right/left click mouse on app icon:
1. If right click mouse: pin menu bar
2. If left click mouse:  app move to main bar and i need tap to icon on main bar

**Comment by underscoreYTREWQ** [comment]
Just switched back to MacOS from Windows, and started using Ice. Absolutely amazing app.

suggestion for the second bar —

Maybe an option to make the bar appear under the Ice icon on the main menu bar? Instead of always opening justifies to the right of the screen.

**Comment by jordanbaird** [comment]
@fleytman Good idea about the pre-release. I was under the impression that Homebrew didn't support doing pre-releases, but I think it should be fine.

**Comment by jordanbaird** [comment]
Beta 5 is up as a pre-release. I went ahead and made pre-releases for the previous betas as well so they're all easier to access.

At this point what I would like to know is how close this feels to being ready for release. If there are some breaking bugs, we'll obviously have to address those, but if it's just down to small inconsistencies or missing features, those can be added or fixed later.

**Comment by lint2015** [comment]
Thanks again for the updates @jordanbaird. With beta 5, my previous issues persisted initially, however instead of the zero-sized Ice Bar, I got a pop up saying something like "Unable to display menu bar items. Try switching spaces."

Happy to report that once I did this, switched back to my original space, and now both the Menu Bar Items section of Preferences and the Ice Bar itself are functioning as intended.

I'll post back if I encounter any serious bugs relating to the Ice Bar.

**Comment by dploeger** [comment]
Beta 5 works fine with the Ice bar for me. However, the "Menu Bar Items" settings look like this:

<img width="672" alt="image" src="https://github.com/jordanbaird/Ice/assets/203453/bdb207ea-fe12-40dc-a059-f9174fb1ef85">

Maybe a permission is missing here?

**Comment by jordanbaird** [comment]
@dploeger What happens if you switch to another space? I've found that sometimes a space seems to become "invalid" for offscreen window capture. Once you switch to another space, you can switch back and it works again.

**Comment by jordanbaird** [comment]
@dploeger Hmm...does the number of question marks correspond to the number of visible items in your menu bar?

**Comment by dploeger** [comment]
@jordanbaird Yes. And the feature actually works. I can move around the question marks and the items in the menu bar are reordered.

**Comment by jonasjelonek** [comment]
With beta 5 I have some issues again with the Ice Bar and a second screen. If I open the Ice Bar on the main screen after Ice's first start, its height is too low so that the icons are cut off. But after a second or two this fixes itself. However, when I open the Ice Bar on my second screen it and the icons within are displayed extremely large (see below). Sometimes, there are also some temporary issues with the icon sizing but they usually resolve after closing and reopening the Ice Bar one multiple times.

![Screenshot 2024-06-24 at 09 52 12](https://github.com/jordanbaird/Ice/assets/52761381/19592aec-4ff9-4707-b523-31ec1cb4757c)

**Comment by sweetppro** [comment]
@jordanbaird beta5 is working great for me. thanks!

**Comment by T-Specht** [comment]
> At this point what I would like to know is how close this feels to being ready for release. If there are some breaking bugs, we'll obviously have to address those, but if it's just down to small inconsistencies or missing features, those can be added or fixed later.

Thank you for your hard work, beta 5 works great for me, too! Also, the workaround for Magnet that you implemented has fixed the issues I was having previously.

**Comment by BryanWall** [comment]
When I tried beta 5 coming from 0.9.0, none of my icons were displayed and all of the sections of "Menu Bar Items" in settings were blank (screenshot attached). I rolled back to 0.9.0 and then it wouldn't work either. I used CleanMyMac to remove Ice and all of its settings. After purging all traces of Ice and reinstalling beta 5, everything is working as expected. Maybe there is an issue converting some settings from 0.9.0 to the latest versions?
<img width="973" alt="Screenshot 2024-06-24 at 1 36 45 PM" src="https://github.com/jordanbaird/Ice/assets/5903057/e3b4ab41-b5a4-4ebe-8c31-fe6793159996">

**Comment by CzarHey** [comment]
Bartender refugee here. The Ice Bar is coming along nicely, though I noticed an issue where Ice doesn't hide the application's menu bar items when using the Ice Bar, even when "Hide application menus when showing menu bar items" is selected. This issue doesn't appear if the Ice Bar is disabled; application menus will be hidden normally in that instance.

https://github.com/jordanbaird/Ice/assets/68244775/78a3ea9b-b7af-4ad7-8391-057c564a5760

**Comment by jordanbaird** [comment]
@gingerbeardman Did the coloring issue ever get fixed for you? The Ice Bar is supposed to be roughly the color of the menu bar on the active screen, and the menu bar items should match the color of the items on that same screen.

**Comment by gingerbeardman** [comment]
I'll try the beta again later today and let you know

**Comment by gingerbeardman** [comment]
> @gingerbeardman Did the coloring issue ever get fixed for you? The Ice Bar is supposed to be roughly the color of the menu bar on the active screen, and the menu bar items should match the color of the items on that same screen.

No, still a problem in light mode:

light mode
- icons are white (would expect them to be dark!)
- bar is roughly the menu bar colour

dark mode
- icons are white
- bar is roughly the menu bar colour

I also see two other problems
1. i have two displays - MBP which is retina and external display (set as main) which is not
   - bar shows up tiny on retina MBP display
   - bar is normal sized on external/main display
1. "unable to display menu bar items. try switching spaces
    - but i do not use spaces at all
    - seems very temperamental whether or not the bar will appear

**Comment by jonasjelonek** [comment]
@jordanbaird I probably noticed a similar issue again. When I change my wallpaper (I do that regularly with an app) and the menu bar color changes, the background of the Ice bar changes properly too, but not the icon color (again). IIRC this was working in the last beta, at least for me.

@gingerbeardman Just out of curiosity, is the color of the menu bar influenced by light/dark mode on your system? For me it isn't, but only depends on the background that is used. I also have auto mode for light/dark activated but my menu bar color only changes when I change the wallpaper.

**Comment by gingerbeardman** [comment]
My menu bar colour changes along with light and dark mode.

**Comment by jordanbaird** [comment]
@gingerbeardman This is how it looks on my system while switching between light and dark mode with a dynamic wallpaper:

https://github.com/jordanbaird/Ice/assets/90936861/a3771495-b567-40e4-8d4e-9ea5cac94aa9

Note that in macOS, the menu bar item color is influenced by the color of the menu bar, not the specific system appearance. They might not always change colors between light and dark mode, since some people might not use dynamic wallpapers. So it isn't an easy check to determine which items to show. The way I've been dealing with it so far is to just capture the current item windows. To be honest, I have no idea why it isn't capturing the right ones for you.

On the "try switching spaces" message:

Menu bar item capture, in my experience, isn't reliable at all on some people's machines. What seems to be happening is that a space can end up in an "invalid" state of sorts and suddenly, none of its menu bar items can be captured. The reason for the "try switching spaces" message is because that's the only way I've been able to figure out to get a space out of this state. If anyone has ideas for this, I'd love to hear them

**Comment by vr8hub** [comment]
Installed beta 5 (on Sonoma 14.4.1, M3 MBP); the background color of the Ice Bar is fixed for me (now light like the main menu), but clicking on any of the icons in the bar still doesn't open the associated app. It does lighten the icon for a few seconds then goes back to normal, but the app doesn't open. True for every icon in the bar (I have eight).

**Comment by jordanbaird** [comment]
@vr8hub Good to know. Still work to be done on that front, I suppose.

**Comment by apotenza92** [comment]
Beta 5 works super well for me thanks so much for your efforts @jordanbaird! One minor suggestion I'd like is a downwards facing chevron as an icon! Would compliment the Bar well.

**Comment by gingerbeardman** [comment]
> dynamic wallpapers

I don't use dynamic wallpaper. 

And I will record a video of how the transition from light to dark is handled for me

**Comment by fleytman** [comment]
Thank you, overall it has become more stable, I almost completely switched to ice bar!According to the latest beta and compared to ibar.
1. Left mouse click now opens the app menu, thank you. But it doesn't work in full screen mode (fn+f)
2. The right click didn't appear. I know that ibar doesn't have it either. Whether there is any organichenie on PCM?
3. If in ice settings that shows iknoku of record of the screen. In ibar, the icon appears less frequently (or does not appear at all)
4. New elements are added to autohide, although it is more logical to add them to the second panel
5. As I mentioned earlier, in ibar you can change the size of icons, it is worth it [to implement] (https://www.jessesquires.com/blog/2023/12/16/macbook-notch-and-menu-bar-fixes/)
6. The znotch function is also not bad to add, sometimes you need to place more icons in front of your eyes

**Comment by vr8hub** [comment]
I don't know if anything is feasible, but…
Fantastical shows text for your next appointment next to its icon in the menubar when the appt gets close (within an hour? 90 minutes?). This of course pushes everything else to the left, which means (for me) at least one and sometimes two icons next to the notch get pushed underneath it and are no longer visible. The Ice Bar doesn't help in that circumstance, since the icons aren't in the hidden section. It would be nice™ if Ice could include those in the Ice Bar for as long as Fantastical is taking up extra room.
As I said, maybe not feasible, but I thought I'd ask just in case.

**Comment by Xytronix** [comment]
Bug: when closing an app and reopening it, ice doesn’t remember its position. Instead it is added to the menu bar.

**Comment by gingerbeardman** [comment]
> On the "try switching spaces" message:

Switching spaces still doesn't get the icons to appear for me, I still get the message.

I use an external display, portrait, that is sometimes retina sometimes not.

Right now I'm unable to see the ice bar, even restarting the app doesn't help.

**Comment by Xytronix** [comment]
<img width="759" alt="image" src="https://github.com/jordanbaird/Ice/assets/32957125/6346e874-6202-4f9f-8107-567e6c5bc680">
Icon doesn't show up under Ice for a reason. It gets removed every time I terminate the app as well.

**Comment by jordanbaird** [comment]
@Xytronix what icon are you referring to?

**Comment by Xytronix** [comment]
@jordanbaird the top left icon next to the dots

**Comment by BlackWoxs** [comment]
> @jordanbaird Yes. And the feature actually works. I can move around the question marks and the items in the menu bar are reordered.

I see a similar issue of the ? marks on my always hidden section - I discovered if I drag the ? to another section in the settings and then back again to its original menu the display is corrected

**Comment by jordanbaird** [comment]
There's a new beta release up [here](https://github.com/jordanbaird/Ice/releases/tag/0.10.0-beta.6). Let's move any discussion about bugs to #192. I'm going to close this issue, since notch support is now implemented.
